### ISSUE TITLE
xl.trax.trackdb fails to save database

### ISSUE THREAD

**Issue by genodeftest** [exaile/exaile#386]
### Steps to Reproduce (for bugs)
1. use the new 4.0.0-beta1 installer
2. start Exaile
3. have a look at the log file

### Expected Behavior
No tracebacks. No errors.

### Current Behavior
Upon opening the database (with 'c' flag for creating it if not present), I get a traceback:
```
18:56:52 INFO    : Loading Exaile 4.0.0beta1... (xl.main)
18:56:52 INFO    : Using Python 2.7.13 (xl.version)
18:56:52 INFO    : Using PyGObject 3.24.1 (xl.version)
18:56:52 INFO    : Loading settings... (xl.main)
18:56:52 INFO    : Using Locale de_DE cp1252 (xl.version)
18:56:52 INFO    : Using Mutagen 1.37 (xl.version)
18:56:54 INFO    : Using GStreamer 1.10.4 (xl.version)
18:56:54 INFO    : Reconfiguring crossfading (xl.player.gst.engine [player])
18:56:54 INFO    : Crossfade: disabled (xl.player.gst.engine [player])
18:56:54 INFO    : Using GTK+ 3.22.15 (xl.version)
18:56:55 INFO    : Loading plugins... (xl.main)
18:56:56 INFO    : Loading collection... (xl.main)
18:56:56 ERROR   : Failed to open music DB. (xl.trax.trackdb)
Traceback (most recent call last):
  File "C:/msys32/home/IEUser/exaile/tools/installer/_inst/usr/lib/exaile/xl/trax/trackdb.py", line 186, in load_from_location
  File "C:/msys32/home/IEUser/exaile/tools/installer/_build_root/mingw32/lib/python2.7/shelve.py", line 243, in open
  File "C:/msys32/home/IEUser/exaile/tools/installer/_build_root/mingw32/lib/python2.7/shelve.py", line 227, in __init__
  File "C:/msys32/home/IEUser/exaile/tools/installer/_build_root/mingw32/lib/python2.7/anydbm.py", line 82, in open
error: db type could not be determined
```

### Possible Solution
No clue.

### Environment
* Operating System and version: Windows 10
* Exaile Version: 4.0.0-beta1 from [official installer](https://github.com/exaile/exaile/releases/tag/4.0.0-beta1)

**Comment by genodeftest** [comment]
This issue has been reported to python upstream developers: https://bugs.python.org/issue23174. They claim it is not a bug, but the behavior differs from the docs:

https://docs.python.org/2.7/library/anydbm.html?highlight=anydbm.open#anydbm.open

>  anydbm.open(filename[, flag[, mode]])
> 
>    Open the database file filename and return a corresponding object.
> 
>    If the database file already exists, the `whichdb` module is used to determine its type and the appropriate module is used; if it does not exist, the first module listed above that can be imported is used.

### Possible Solution
If the file does not exist yet, we have to specify our own database type.

**Comment by virtuald** [comment]
I can reproduce this on my VM, I'll see what's going on. Maybe there's a library I'm missing?

**Comment by virtuald** [comment]
My msys2 install picks gdbm when I use the shelf module, so if I run the following:

```
import gdbm
db = gdbm.open('test.db', 'c')
```

This fails with:

```
Traceback (most recent call last):
  File "foo.py", line 2, in <module>
    db = gdbm.open('test.db', 'c')
gdbm.error: (22, 'Invalid argument')
```

Which creates an empty file, which then triggers https://bugs.python.org/issue23174. Something weird is happening here... maybe the python build for msys2 is bad somehow? I'll have to think about the best way to debug this.

Well, I ran it on my Linux box, and it works fine, and the result of running `file` on it is:
```
test.db: GNU dbm 1.x or ndbm database, little endian, 64-bit
```
Interestingly enough, this isn't what python picks for shelf, it picks:
```
music.db: Berkeley DB (Hash, version 9, native byte-order)
```

Can someone else with an msys2 installation try that on their machine and see what happens? I have gdbm 1.11-3 installed.. no, wait, I have two of them?

```
$ pacman -Ss gdbm
mingw32/mingw-w64-i686-gdbm 1.12-1 [installed]
    GNU database library (mingw-w64)
mingw64/mingw-w64-x86_64-gdbm 1.12-1
    GNU database library (mingw-w64)
msys/gdbm 1.11-3 (Database) [installed]
    GNU database library
msys/libgdbm 1.11-3 (libraries) [installed]
    GNU database library
msys/libgdbm-devel 1.11-3 (development)
    libgdbm headers and librarie
```

**Comment by sjohannes** [comment]
Yeah I can confirm all of that. My music.db is carried over from the pygi-aio times so it's working fine, but if I start with no db I get the problem.

shelve defaults to using dbhash (which uses bsddb) on Linux and Windows-native, but apparently uses gdbm on Windows-msys2 which has that bug.

(Off topic, but this could be a huge issue when moving to Python 3 because dbhash and bsddb have been removed there and the existing music.db can't be read.)

> I have gdbm 1.11-3 installed.. no, wait, I have two of them?

Only the mingw{32,64} ones are relevant. The msys one is for building msys packages or something, I don't really remember.

**Comment by genodeftest** [comment]
> (Off topic, but this could be a huge issue when moving to Python 3 because dbhash and bsddb have been removed there and the existing music.db can't be read.)

Yeah, that might be hard. It's not forgotten but tracked in #246.

**Comment by genodeftest** [comment]
> Can someone else with an msys2 installation […]

No msys2 here, sorry.

**Comment by sjohannes** [comment]
>> (Off topic, but this could be a huge issue when moving to Python 3 because dbhash and bsddb have been removed there and the existing music.db can't be read.)
>
> Yeah, that might be hard. It's not forgotten but tracked in #246.

(I edited the comment at the same time you posted.) Looks like there's an external bsddb3 module that we can use; just need to be specific when opening the shelf file: `shelve.BsdDbShelf(bsddb3.hashopen(...))`.

**Comment by sjohannes** [comment]
Oh hey, that bsddb3 trick actually also works on msys2. @virtuald could you try that?

**Comment by genodeftest** [comment]
> Oh hey, that bsddb3 trick actually also works on msys2. @virtuald could you try that?

This might break:

https://docs.python.org/2.7/faq/library.html?highlight=bsddb#i-tried-to-open-berkeley-db-file-but-bsddb-produces-bsddb-error-22-invalid-argument-help-how-can-i-restore-my-data

**Comment by virtuald** [comment]
Hm, looking at what we're doing... I don't quite understand why we're using shelf. It seems like shelf is... a disk backed database that pretends to be a dictionary? But we're not using it that way, we're using it strictly as a file format.

It seems like it would be better to use pickle by itself, which would help with the python 3 migration since we wouldn't need to worry about bsddb. There's probably a downside though, right?

**Comment by genodeftest** [comment]
> It seems like it would be better to use pickle by itself, which would help with the python 3 migration since we wouldn't need to worry about bsddb. There's probably a downside though, right?

Since we already used bsddb to create the database, isn't it saved in bsddb format? If yes, we'll have to ship bsddb anyway…

**Comment by sjohannes** [comment]
> Hm, looking at what we're doing... I don't quite understand why we're using shelf. It seems like shelf is... a disk backed database that pretends to be a dictionary? But we're not using it that way, we're using it strictly as a file format.

My understanding: If you have a list that you want to dump into a file using pickle, you have to actually pickle the whole list and rewrite the whole file, regardless of how many items have changed. With a shelve, you can just update items that have changed.

**Comment by virtuald** [comment]
Hm. I had missed that.

I think I'll try to build a new version of gdbm (it's at 1.13, and the version in msys is 1.12) and see if the issue goes away.

**Comment by virtuald** [comment]
Never got around to doing that, but on IRC we came to the conclusion that we should try to force bsddb. To that end, I've extended the db explorer tool in ecc896b3b0eff2eedce98cc6562fa20affe0ae70 to allow us to easily convert databases from one format to another for experimentation purposes.

**Comment by virtuald** [comment]
Fixed in #399
